### DISCUSSION TITLE
chapter_backtracking/summary/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#562] [Announcements]
# chapter_backtracking/summary/

动画图解、一键运行的数据结构与算法教程

https://www.hello-algo.com/chapter_backtracking/summary/

**Comment by panda-on** [comment]
13章完成，撒花🎉

**Reply by sunshinesDL** [reply]
K大，小结里第2条：“**回溯通过深度优先搜索来尝试各种选择，当遇到不满足约束条件的情况时，则撤销上一步的选择，退回到之前的状态，并继续尝试其他选择”，其中 “当遇到不满足约束条件的情况时，则撤销上一步的选择**” 是不是不太对？
我认为这里应该是 “**在把当前状态后续的所有可能性遍历完成后（对应下面递归函数的返回），则撤销这一步的选择（对应撤销这轮循环做出的选择），以回退到执行这一步之前的状态**”。 
这样理解恰当吗？@krahets  望 K大 及评论区大佬指正。 😁

**Comment by ShiinaMahiru-repo** [comment]
完结撒花， 这一章太痛苦了

**Reply by SoftwareEngineerPalace** [reply]
k 神讲得也是太好，如果没有他，你可能会再痛苦 n 倍

**Reply by qw124341789** [reply]
理解这些不是很难，难的是把这些转换成C代码

**Reply by nors1993** [reply]
确实是啊，逻辑上能理解，但是放到代码里就不太好理解了。。。。

**Comment by lizujin-magictavern-com** [comment]
mark

**Comment by ghost** [comment]
秒啊

**Reply by ghost** [reply]
太妙了！！感谢K大佬

**Comment by Magicperformer** [comment]
大佬你好，我想请问一下分治和回溯是否比较相像呢？感觉都是解决算法问题的一个思路，很难完全分清。同时分治问题当中介绍的汉诺塔问题也更像不同状态之间的转换，也非常像回溯的解题思路。

**Reply by JianhCode** [reply]
我认为，回溯算法的一大特征就是剪枝，而分治算法中没有剪枝操作。同时，分治是通过递归将问题拆解为最小问题(例如在汉诺塔问题的递归树中的f(1))，然后通过回归从下往上将递归树的问题进行求解，进而得到原问题的解，旨在以小化大。回溯算法则类似于前序遍历所有选择所构成的树，尝试所有解，然后通过剪枝和回溯来判断当前选择是否符合条件，最终以获得所有符合目标的解，旨在穷举的过程中找到符合条件的解。

**Reply by krahets** [reply]
是比较像，因为它们都是基于分解子问题来解决问题的。不过两者的侧重点不同，分治侧重于划分与合并，回溯侧重于搜索与剪枝

**Comment by wangx77** [comment]
😵

**Comment by zhemintia** [comment]
debug 小结最后一行的“找处在”应为“找出在”

**Reply by jiuerya7** [reply]
人家说的是“找出处在同一行”

**Comment by Unsichtbare1** [comment]
太棒了，照猫画虎自己写了个解决数独的算法。成就感满满
